### PR TITLE
chore: create variables map protocol for use in core

### PR DESCRIPTION
**Pull request by frodehk**
ECALC-1664

## Have you remembered and considered?

- [ ] I have remembered to update documentation
- [ ] I have remembered to update manual changelog (`docs/docs/changelog/next.md`)
- [ ] I have remembered to update migration guide (`docs/docs/migration_guides/`)
- [x] I have committed with `BREAKING:` in footer or `!` in header, if breaking
- [ ] I have added tests (if not, comment why)
- [x] I have used conventional commits syntax (if you squash, make sure that conventional commit is used)
- [x] I have included the Jira issue ID somewhere in the commit body (`ECALC-XXXX`)

## Why is this pull request needed?

Introduce protocol with get-methods for VariablesMap, to reduce core-dependency and increase flexibility.

## What does this pull request change?

Implement VariablesMapService (Protocol) in variables.

## Issues related to this change:
https://equinor-ecalc.atlassian.net/browse/ECALC-1664?atlOrigin=eyJpIjoiNjg4YTVjNjZiZGNiNDE2M2JmNmJjNDcyZTg5MTJjNjciLCJwIjoiaiJ9

### CODE REVIEW

File: src/libecalc/common/variables.py [hunk 1]
```diff
@@ -65,3 +66,45 @@ def get_subset_for_timestep(self, current_timestep: datetime) -> VariablesMap:
...
+    def get_time_vector(self) -> [List[datetime]]: ...
+
+    @abc.abstractmethod
+    def get_variables(self) -> Dict[str, List[Annotated[float, Field(allow_inf_nan=False)]]]: ...
+
+    @abc.abstractmethod
+    def get_period(self) -> Period: ...
+
+    @abc.abstractmethod
+    def get_subset(self, start_index: int, end_index: int) -> VariablesMap: ...
```
Comment (jsolaas) [inline src/libecalc/common/variables.py:101]:
```suggestion
    def get_subset(self, start_index: int, end_index: int) -> VariablesMapService: ...
```

File: src/libecalc/common/variables.py [hunk 2]
```diff
@@ -65,3 +66,45 @@ def get_subset_for_timestep(self, current_timestep: datetime) -> VariablesMap:
...
+
+    @abc.abstractmethod
+    def get_subset_from_period(self) -> VariablesMap: ...
```
Comment (jsolaas) [inline src/libecalc/common/variables.py:104]:
```suggestion
    def get_subset_from_period(self) -> VariablesMapService: ...
```

File: src/libecalc/common/variables.py [hunk 3]
```diff
@@ -65,3 +66,45 @@ def get_subset_for_timestep(self, current_timestep: datetime) -> VariablesMap:
...
+
+    @abc.abstractmethod
+    def get_subset_for_timestep(self) -> VariablesMap: ...
```
Comment (jsolaas) [inline src/libecalc/common/variables.py:107]:
```suggestion
    def get_subset_for_timestep(self) -> VariablesMapService: ...
```

File: src/libecalc/common/variables.py [hunk 4]
```diff
@@ -65,3 +66,45 @@ def get_subset_for_timestep(self, current_timestep: datetime) -> VariablesMap:
...
+
+    @abc.abstractmethod
+    def get_variables_map(self) -> VariablesMap: ...
```
Comment (jsolaas) [inline src/libecalc/common/variables.py:110]:
```suggestion
    def get_variables_map(self) -> VariablesMapService: ...
```

File: src/libecalc/common/variables.py [hunk 5]
```diff
@@ -65,3 +66,45 @@ def get_subset_for_timestep(self, current_timestep: datetime) -> VariablesMap:
...
+
+
+T = TypeVar(
+    "T",
+    bound=Union[List[datetime], Dict[str, List[Annotated[float, Field(allow_inf_nan=False)]]], Period, VariablesMap],
+    covariant=True,
+)
+
+
+class VariablesMapService(Protocol[T]):
```
Comment (jsolaas) [inline src/libecalc/common/variables.py:90]:
```suggestion
class VariablesMapService(Protocol):
```

File: src/libecalc/common/variables.py [hunk 6]
```diff
@@ -65,3 +66,45 @@ def get_subset_for_timestep(self, current_timestep: datetime) -> VariablesMap:
...
+    @abc.abstractmethod
+    def get_time_vector(self) -> [List[datetime]]: ...
+
+    @abc.abstractmethod
+    def get_variables(self) -> Dict[str, List[Annotated[float, Field(allow_inf_nan=False)]]]: ...
```
Comment (jsolaas) [inline src/libecalc/common/variables.py:95]:
```suggestion
    def get_variables(self) -> Dict[str, List[float]]: ...
```

I don't think the annotations does anything here?

File: src/libecalc/common/variables.py [hunk 7]
```diff
@@ -65,3 +66,45 @@ def get_subset_for_timestep(self, current_timestep: datetime) -> VariablesMap:
...
+
+    def get_variables_map(self):
+        return self
+
+
+T = TypeVar(
+    "T",
+    bound=Union[List[datetime], Dict[str, List[Annotated[float, Field(allow_inf_nan=False)]]], Period, VariablesMap],
+    covariant=True,
+)
```
Comment (jsolaas) [inline src/libecalc/common/variables.py:87]: I don't think this does anything. Using a TypeVar would require _some_ input to variables map that decided what the type is, and that the `T` was used as a return type in the methods. Unless I misunderstood something.

**Review by jsolaas — COMMENTED** [review]
Looks like a good start! I think we should look into redefining the interface also, which would require more work.

If for example we made an evaluate function in `VariablesMapService` we could make evaluating expressions the responsibility of the VariablesService. That way we wouldn't have to expose as much of the internals. 

```
def evaluate(expression_or_variable_or_temporal_expression) -> List[float]: ...
```

An alternative is to get rid of the whole VariablesMap also, and instead let the expressions be able to evaluate itself, I think....

**Comment by TeeeJay** [comment]
> Looks like a good start! I think we should look into redefining the interface also, which would require more work.
> 
> If for example we made an evaluate function in `VariablesMapService` we could make evaluating expressions the responsibility of the VariablesService. That way we wouldn't have to expose as much of the internals.
> 
> ```
> def evaluate(expression_or_variable_or_temporal_expression) -> List[float]: ...
> ```
> 
> An alternative is to get rid of the whole VariablesMap also, and instead let the expressions be able to evaluate itself, I think....

Agree, very good start indeed. I guess the example would also require to send in the relevant timesteps/vector, as that information is only where you need to evaluate the expression etc atm?

We can potentially also rename the "VariablesMap/Service" to something that better fits its responsibility in that case, such as `Evaluator` since it can evaluate "a lot of things" to (list of) floats ... ?

**Review by jsolaas — APPROVED** [review]

**Pull request merged**